test/client.js

```diff
@@ -389,5 +389,32 @@ describe('MqttClient', function () {
         })
       })
     })
+
+    it('should resubscribe even if disconnect is before suback', function (done) {
+      var client = mqtt.connect(Object.assign({ reconnectPeriod: 100 }, config))
+      var subscribeCount = 0
+      var connectCount = 0
+
+      server.removeAllListeners('client')
+      server.on('client', function (serverClient) {
+        serverClient.on('connect', function () {
+          connectCount++
+          serverClient.connack({returnCode: 0})
+        })
+
+        serverClient.on('subscribe', function () {
+          subscribeCount++
+          if (subscribeCount === 1) {
+            client.stream.end()
+          } else {
+            connectCount.should.equal(2)
+            subscribeCount.should.equal(2)
+            client.end(true, done)
+          }
+        })
+      })
+
+      client.subscribe('hello')
+    })
```

Comment:
I fixed the test state leakage issue in e9c78db. The fix is somewhat gross, but at least it's self contained.